test/FunctionalTests/Microsoft.OData.Core.Tests/ODataMessageWriterTests.cs

```diff
@@ -302,6 +446,88 @@ public void WhenInjectingStreamBasedJsonWriterFactory_ThrowException_IfFactoryRe
             writePropertyAction.Throws<ODataException>(Strings.ODataMessageWriter_StreamBasedJsonWriterFactory_ReturnedNull(Encoding.UTF8.WebName, false));
         }
 
+        [Fact]
+        public async Task SupportsStreamBasedJsonWriterAsync()
+        {
+            ODataMessageWriterSettings settings = new ODataMessageWriterSettings();
+
+            using MemoryStream stream = new MemoryStream();
+            InMemoryMessage request = new InMemoryMessage() { Stream = stream };
+
+            IContainerBuilder containerBuilder = new Test.OData.DependencyInjection.TestContainerBuilder();
+            containerBuilder.AddDefaultODataServices();
+            containerBuilder.AddService<IStreamBasedJsonWriterFactory>(
+                ServiceLifetime.Singleton, sp => DefaultStreamBasedJsonWriterFactory.Default);
+            request.Container = containerBuilder.BuildContainer();
+
+            IStreamBasedJsonWriterFactory factory = request.Container.GetService<IStreamBasedJsonWriterFactory>();
+            Assert.IsType<DefaultStreamBasedJsonWriterFactory>(factory);
+
+            settings.ODataUri.ServiceRoot = new Uri("http://host/service");
+            settings.SetContentType(ODataFormat.Json);
+            EdmModel model = new EdmModel();
+            using ODataMessageWriter writer = new ODataMessageWriter((IODataRequestMessage)request, settings, model);
+            await writer.WritePropertyAsync(new ODataProperty()
+            {
+                Name = "Name",
+                Value = "This is a test ия"
+            });
+
+            request.GetStream().Position = 0;
+            using StreamReader reader = new StreamReader(request.GetStream());
+            string output = await reader.ReadToEndAsync();
+            Assert.Equal("{\"@odata.context\":\"http://host/service/$metadata#Edm.String\",\"value\":\"This is a test \\u0438\\u044F\"}", output);
+        }
+
+        [Theory]
+        [InlineData("utf-8")]
+        [InlineData("utf-16")]
+        [InlineData("utf-16BE")]
+        [InlineData("utf-32")]
+        public async Task WhenInjectingStreamBasedJsonWriterFactoryAsync_CreatesWriterUsingConfiguredEncoding(string encodingCharset)
+        {
+            // Arrange
+            ODataMessageWriterSettings settings = new ODataMessageWriterSettings();

```

Comment:
You could move some of the logic to a helper method and reuse it in this test and the other one above

Reply:
I've moved code to set up the container in a helper method.

Reply:
I've also extended the existing helper methods `WriteAndGetPayload/Async` to handle the test cases I added. It did involve a bit of a parameter bloat.